misk-policy/src/main/kotlin/misk/policy/opa/OpaPolicyEngine.kt

```diff
@@ -0,0 +1,93 @@
+package misk.policy.opa
+
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.Types
+import okhttp3.ResponseBody
+import javax.inject.Inject
+import javax.inject.Named
+
+/**
+ * Support for the Open Policy Engine (OPA).
+ * OPA provides a means to decouple policy from business logic. The resulting query response may
+ * have arbitrary shapes.
+ */
+class OpaPolicyEngine @Inject constructor(
+  private val opaApi: OpaApi,
+  @Named("opa-moshi") private val moshi: Moshi
+) {
+
+  /**
+   * Evaluate / Query a document with given input of shape T.
+   * This will connect to OPA via a retrofit interface and perform a /v1/data/{document} POST.
+   *
+   * @throws RuntimeException if the request to OPA failed or the response shape didn't match R.
+   * @return Response shape R from OPA.
+   */
+  fun <T, R> evaluateInternal(
+    document: String,
+    input: T,
+    inputType: Class<T>,
+    returnType: Class<R>
+  ): R {
+    val inputAdapter = moshi.adapter<Request<T>>(
+      Types.newParameterizedType(Request::class.java, inputType)
+    )
+    val inputString = inputAdapter.toJson(Request(input))
+    val response = queryOpa(document, inputString)
+    return parseResponse(returnType, response)
+  }
+
+  inline fun <reified T, reified R> evaluate(document: String, input: T): R {
+    return evaluateInternal(document, input, T::class.java, R::class.java)
+  }
+
+  /**
+   * Evaluate / Query a document with no additional input.
+   * This will connect to OPA via a retrofit interface and perform a /v1/data/{document} POST.
+   *
+   * @throws RuntimeException if the request to OPA failed or the response shape didn't match R.
+   * @return Response shape R from OPA.
+   */
+  fun <R> evaluateInternal(document: String, returnType: Class<R>): R {
```

Comment:
can be `private`

Reply:
actually I tried to make that happen, but because the helper routines are inlined they would be exposing these private functions. I think I'm missing something here on how to hide the internals away properly using generics / inlines